File: app/src/main/java/com/github/willjgriff/ethereumwallet/ui/settings/adapter/ChangeAddressAdapter.java
package com.github.willjgriff.ethereumwallet.ui.settings.adapter;

import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.ViewGroup;

import com.github.wiljgriff.ethereumwallet.ethereum.account.delegates.AccountDelegate;
import com.github.willjgriff.ethereumwallet.R;
import com.github.willjgriff.ethereumwallet.ui.settings.adapter.ChangeAddressItemViewHolder.ChangeAddressItemListener;

import java.util.List;

import static com.github.willjgriff.ethereumwallet.ui.settings.adapter.ChangeAddressAdapter.ChangeAddressType.HEADER;
import static com.github.willjgriff.ethereumwallet.ui.settings.adapter.ChangeAddressAdapter.ChangeAddressType.ITEM;

/**
 * Created by Will on 28/02/2017.
 */

public class ChangeAddressAdapter extends RecyclerView.Adapter<RecyclerView.ViewHolder> {

	private static final int HEADER_OFFSET = 1;
	private List<AccountDelegate> mAccounts;
	private ChangeAddressItemListener mChangeAddressItemListener;

	public ChangeAddressAdapter(ChangeAddressItemListener changeAddressItemListener) {
		mChangeAddressItemListener = changeAddressItemListener;
	}

	public void setAccounts(List<AccountDelegate> accounts) {
		mAccounts = accounts;
	}

	enum ChangeAddressType {
		HEADER,
		ITEM
	}

	@Override
	public RecyclerView.ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
		LayoutInflater layoutInflater = LayoutInflater.from(parent.getContext());
		switch (ChangeAddressType.values()[viewType]) {
			case HEADER:
				return new ChangeAddressHeaderViewHolder(
					layoutInflater.inflate(R.layout.view_change_address_header, parent, false));
			case ITEM:
				return new ChangeAddressItemViewHolder(
					layoutInflater.inflate(R.layout.view_change_address_item, parent, false), mChangeAddressItemListener);
		}
		return null;
	}

	@Override
	public void onBindViewHolder(RecyclerView.ViewHolder holder, int position) {
		if (holder instanceof ChangeAddressItemViewHolder) {
			((ChangeAddressItemViewHolder) holder).bind(mAccounts.get(position - HEADER_OFFSET));
		}
	}

	@Override
	public int getItemCount() {
		return mAccounts.size() + HEADER_OFFSET;
	}

	@Override
	public int getItemViewType(int position) {
		if (position == 0) {
			return HEADER.ordinal();
		}
		return ITEM.ordinal();
	}
}

File: app/src/main/kotlin/com/github/willjgriff/ethereumwallet/ui/settings/list/ChangeAddressAdapter.kt
package com.github.willjgriff.ethereumwallet.ui.settings.list

import android.support.v7.widget.RecyclerView
import android.view.ViewGroup
import com.github.willjgriff.ethereumwallet.R
import com.github.willjgriff.ethereumwallet.ethereum.address.model.DomainAddress
import com.github.willjgriff.ethereumwallet.ui.settings.list.ChangeAddressAdapter.ChangeAddressType.HEADER
import com.github.willjgriff.ethereumwallet.ui.settings.list.ChangeAddressAdapter.ChangeAddressType.ITEM
import com.github.willjgriff.ethereumwallet.ui.settings.list.ChangeAddressItemViewHolder.ChangeAddressItemListener
import com.github.willjgriff.ethereumwallet.ui.utils.inflate

/**
 * Created by williamgriffiths on 04/05/2017.
 */
class ChangeAddressAdapter(private val changeAddressItemListener: ChangeAddressItemListener)
    : RecyclerView.Adapter<RecyclerView.ViewHolder>() {

    private val HEADER_OFFSET = 1

    var accounts: List<DomainAddress> = listOf()
        set(value) {
            field = value
            notifyDataSetChanged()
        }

    internal enum class ChangeAddressType {
        HEADER,
        ITEM
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecyclerView.ViewHolder? {
        when (ChangeAddressType.values()[viewType]) {
            HEADER -> return ChangeAddressHeaderViewHolder(parent.inflate(R.layout.view_change_address_header))
            ITEM -> return ChangeAddressItemViewHolder(parent.inflate(R.layout.view_change_address_item), changeAddressItemListener)
        }
    }

    override fun onBindViewHolder(holder: RecyclerView.ViewHolder, position: Int) {
        if (holder is ChangeAddressItemViewHolder) {
            holder.bind(accounts[position - HEADER_OFFSET])
        }
    }

    override fun getItemCount(): Int {
        return accounts.size + HEADER_OFFSET
    }

    override fun getItemViewType(position: Int): Int =
            when (position) {
                0 -> HEADER.ordinal
                else -> ITEM.ordinal
            }
}